ocaml/libs/http-lib/http.ml

```diff
@@ -934,81 +934,67 @@ module Url = struct
-            x |> sub_before ':'
-        with Not_found -> (* <hostname|ipv4-literal>:... *)
-                          x
-      )
-    in
-    (* <hostname|ipv4-literal> *)
-    let port x =
-      let port_part =
-        try x |> sub_after ']' |> sub_after ':'
-        with Not_found -> (
-          try (* ...]:port *)
-              x |> sub_after ']'
-          with Not_found -> (
-            try (* ...] *)
-                x |> sub_after ':'
-            with Not_found -> (* ...:port *)
-                              ""
-          )
-        )
+    debug "%s: %s" __FUNCTION__ url ;
+    let fail fmt = Printf.ksprintf failwith fmt in
+    let query = function k, v :: _ -> (k, v) | k, [] -> (k, "") in
+    try
+      let uri = Uri.of_string url in
+      debug "%s: %s" __FUNCTION__ (Uri.to_string uri) ;
+      let data =
+        {
+          uri= (match Uri.path uri with "" -> "/" | path -> path)
+        ; query_params= Uri.query uri |> List.map query
+        }
       in
-      (* no port *)
-      try Some (int_of_string port_part) with _ -> None
-    in
-    let uname_password_host_port x =
-      match Astring.String.cuts ~sep:"@" x with
-      | [_] ->
-          (None, host x, port x)
-      | [uname_password; host_port] -> (
-        match Astring.String.cuts ~sep:":" uname_password with
-        | [uname; password] ->
-            (Some (Basic (uname, password)), host host_port, port host_port)
-        | _ ->
-            failwith
-              (Printf.sprintf "Failed to parse authentication substring: %s"
-                 uname_password
-              )
-      )
+      match Uri.scheme uri with
+      | Some "http" ->
+          let scheme =
+            Http
+              {
+                host= Uri.host uri |> Option.get
+              ; auth=
+                  ( match (Uri.user uri, Uri.password uri) with
+                  | Some user, Some pw ->
+                      Some (Basic (user, pw))
+                  | Some user, None ->
+                      Some (Basic (user, ""))
+                  | _ ->
+                      None
+                  )
+              ; port= Uri.port uri
+              ; ssl= false
+              }
+          in
+          (scheme, data)
+      | Some "https" ->
+          let scheme =
+            Http
+              {
+                host= Uri.host uri |> Option.get
+              ; auth=
+                  ( match (Uri.user uri, Uri.password uri) with
+                  | Some user, Some pw ->
+                      debug "%s: user=%s pw=%s" __FUNCTION__ user pw ;
+                      Some (Basic (user, pw))
+                  | Some user, None ->
+                      debug "%s: user=%s" __FUNCTION__ user ;
+                      Some (Basic (user, ""))
+                  | _ ->
+                      debug "%s: user=none" __FUNCTION__ ;
+                      None
+                  )
+              ; port= Uri.port uri
+              ; ssl= true
+              }
```

Comment:
This is almost the same as the `"http"` case above. The only difference is the `ssl` key and additional logging. Can we factor it out?